Imx943 change irqsteer driver as native driver

biwenli opened the pull request:
Change irqsteer driver as a zephyr native driver.

There are two irqsteer drivers in mcux sdk(mcux-sdk-ng/drivers/irqsteer and mcux-sdk-ng/drivers/irqsteer_1) and the APIs are not compatible. mcux-sdk-ng/drivers/irqsteer_1 just likes a native irqsteer driver(Refer to the linux irqsteer driver: drivers/irqchip/irq-imx-irqsteer.c). So change it as a native driver directly.

biwenli committed:
irqsteer-master: Introduce priority attribute

This adds priority attribute to irqsteer in order to
support multi level interrupts.

We need to make this change atomic so that we don't break
the build.

So, we introduce a new property named 'priority' to
`interrupt-cells` and we increase the number of cells.
Then we update all instances of interrupts
using irqsteer (for imx95, imx8m, imx8qm, imx8qxp and also
update the overlay(imx8mp_evk_mimx8ml8_adsp) for tests.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
samples: drivers: uart_aysnc_api: imx943_evk: m33: add config

Add config for imx943_evk_mimx94398_m33
- Use the sample to test the irqsteer driver

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
tests: arch: arm_custom_interrupt: exclude imx943

Exclude imx943,
  - multi level interrupts feature is enabled defaultly,
    so irqsteer1 driver is enabled, then the test case
    report the build issue:
    multiple definition of z_soc_irq_enable,
    tests/arch/arm/arm_custom_interrupt/src/arm_custom_interrupt.c:47
    first defined here.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
drivers: intc: irqsteer: Drop calling into NXP HAL

Supporting irqsteer using NXP HAL becomes increasingly harder with new
SoCs.

For example now there are two incompatible HAL drivers for IRQ steer
(mcux-sdk-ng/drivers/irqsteer and mcux-sdk-ng/drivers/irqsteer_1).

In order to avoid overcomplicating code and better scaling code for
newer SoCs just drop using the NXP HAL and implement an IRQ Steer native
Zephyr driver

Use irqsteer node of imx943 as example.

New features:
- Support multiple irqsteer instances.
- Indroduce new properties(nxp,irq-offset, nxp,num-irqs).

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
soc: nxp: imx943: m33: enable multi level interrupts

This commit enables multi level interrupts:
- Enable multi level interrupts
  imx943 is a two level interrupts system and
  some interrupts(eg. edma interrupts) depend on irqsteer.
- Increase irq number from 405 to 790
  Actually there are 407(IRQ 0 ~ IRQ 406) interrupts
  from nvic(first level interrupts),
  The second level interrupts are extended by irqsteer,
  it extends 32 x 12 = 384 interrupts,
  So first level interrupts + second level interrupts
  = 407 + 384 = 791(IRQ 0 ~ IRQ 790) interrupts

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
dts: arm: nxp: imx943: m7_0: add irqsteer node

Add irqsteer node for m7_0.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
dts: arm: nxp: imx943: m7_1: add irqsteer node

Add irqsteer node for m7_1.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
dts: arm: nxp: imx94x: add edma2 node

Add edma2 node for lpuart3.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
boards: nxp: imx943_evk: m33: enable edma2

Enable edma2 defaultly.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

LaurentiuM1234 commented on dts/bindings/interrupt-controller/nxp,irqsteer-intc.yaml:
```diff
@@ -7,3 +7,25 @@ include: [base.yaml]
+      So for irqsteer instance 0, irq offset is 300.
+      For irqsteer instance 1, irq offset is 501.
+
+  irqs_num:
```
I'd suggest changing to "nxp,num-irqs" to align with Linux's "fsl,num-irqs"

LaurentiuM1234 commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -225,153 +225,331 @@
+/**
+ * @brief Enable specific interrupt in IRQ_STEER
+ */
+static void irqsteer_enable_interrupt(const struct irqsteer_config *cfg, uint32_t irq)
```
suggestion: combine `irqsteer_enable_interrupt` and `irqsteer_disable_interrupt` into a single function: `irqsteer_enable_disable_interrupt` (or something like that) that takes a bool as the third parameter: `enabled`.

LaurentiuM1234 commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -225,153 +225,331 @@
+/**
+ * @brief Enable master interrupt output
+ */
+static void irqsteer_enable_master_interrupt(const struct irqsteer_config *cfg, int master_index)
```
same as `irqsteer_enable_interrupt()`

LaurentiuM1234 reviewed (commented):
Some initial comments below.

Most importantly: please don't change anything in the driver unless it's related to the fact that you're dropping the HAL API. This is an important change and by doing so you're making the review harder and easier to miss out the important stuff.

If you really want to make some unrelated changes then please do so in a separate commit with the appropriate justification.

Also, please make sure you don't break bisectability (see comment from binding). Make sure that after you apply each commit, the affected boards will still compile.

biwenli commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -225,153 +225,331 @@
-	.master_index = DT_REG_ADDR(node_id),				\
-	.irq = DT_IRQN(node_id),					\
+/* IRQ_STEER register offsets */
+#define CTRL_STRIDE_OFF(_t, _r) (_t * 4 * _r)
```
> please use tab to align with the other macro values

sure, np.

biwenli commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -225,153 +225,331 @@
 #if defined(CONFIG_XTENSA)
-#define irqstr_l1_irq_enable_raw(irq)     xtensa_irq_enable(XTENSA_IRQ_NUMBER(irq))
-#define irqstr_l1_irq_disable_raw(irq)    xtensa_irq_disable(XTENSA_IRQ_NUMBER(irq))
+#define irqstr_l1_irq_enable_raw(irq)      xtensa_irq_enable(XTENSA_IRQ_NUMBER(irq))
```
> same comment as the `LOG_MODULE_REGISTER` stuff above

ok.

biwenli committed:
dts: arm: nxp: imx95: m7: add more masters

Add more irqsteer masters.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
dts: arm: nxp: imx95: m7: add disp_irqsteer node

Add disp_irqsteer node(this is another irqsteer instance).

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli commented:
hal nxp pr:
https://github.com/zephyrproject-rtos/hal_nxp/pull/637

biwenli commented:
> can you please also add a motivation on changing to native driver instead of using HAL. I'm supporting this kind of changes but we need a strong motivation.

@dbaluta ,
There are two irqsteer drivers in mcux sdk(mcux-sdk-ng/drivers/irqsteer and mcux-sdk-ng/drivers/irqsteer_1) and the APIs are not compatible. mcux-sdk-ng/drivers/irqsteer_1 just likes a native irqsteer driver(Refer to the linux irqsteer driver: drivers/irqchip/irq-imx-irqsteer.c). So @LaurentiuM1234 and I want to change the zephyr irqsteer driver as a native driver.

biwenli commented:
> > @dbaluta ,
> > There are two irqsteer drivers in mcux sdk(mcux-sdk-ng/drivers/irqsteer and mcux-sdk-ng/drivers/irqsteer_1) and the APIs are not compatible. mcux-sdk-ng/drivers/irqsteer_1 just likes a native irqsteer driver(Refer to the linux irqsteer driver: drivers/irqchip/irq-imx-irqsteer.c). So @LaurentiuM1234 and I want to change the zephyr irqsteer driver as a native driver.
> 
> Please add this information in the commit message. PR history and this page will be lost in time :)

sure, np.

dbaluta commented on dts/arm/nxp/nxp_imx94x.dtsi:
```diff
@@ -752,221 +784,221 @@
-&gpio5{
+&gpio5 {
 	pinmux = <&iomuxc_eth0_txd0_gpio_io_gpio5_io0>,
-		<&iomuxc_eth0_txd1_gpio_io_gpio5_io1>,
```
Please don't randomly fix whitespaces in the middle of a patch. You can create a separate cleanup PR if you want .

biwenli replied:
> Please don't randomly fix whitespaces in the middle of a patch. You can create a separate cleanup PR if you want .

Fix compliance checks issues.

dbaluta commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -550,6 +550,78 @@
 				#interrupt-cells = <2>;
```
Please update the commit message with a proper explanation.

dbaluta commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -225,20 +225,36 @@
+/* Register access macros */
+#define IRQSTEER_READ32(base, offset)       sys_read32((base) + (offset))
+#define IRQSTEER_WRITE32(base, offset, val) sys_write32((val), (base) + (offset))
+
```
I would get read read of this macros because capital letter make code harder to read.

biwenli replied:
> I would get read read of this macros because capital letter make code harder to read.

Sure, np. I will update it(irqsteer_read_reg32).

dbaluta commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -225,20 +225,36 @@
 
-LOG_MODULE_REGISTER(nxp_irqstr);
+LOG_MODULE_REGISTER(nxp_irqsteer);
 
```
Is there a reason for this change? All the prefixes in this file are still `nxp_irqster`

dbaluta commented on modules/hal_nxp/mcux/mcux-sdk-ng/drivers/drivers.cmake:
```diff
@@ -130,7 +130,6 @@ set_variable_ifdef(CONFIG_MCUX_XBARB            CONFIG_MCUX_COMPONENT_driver.xba
 set_variable_ifdef(CONFIG_QDEC_MCUX             CONFIG_MCUX_COMPONENT_driver.enc)
```
This could and should be integrated in the commit that drops the HAL NXP calls.

biwenli commented on dts/arm/nxp/nxp_imx943_m33.dtsi:
```diff
@@ -14,7 +14,7 @@
 
-		cpu0:cpu@0 {
+		cpu0: cpu@0 {
 			device_type = "cpu";
```
> Please don't add random cleanup fixes in the middle of a patch. If you want to do cleanups please add them at the beginning of the PR or create a totally new PR.

Okay, I will add them at the begginning of the PR.

biwenli commented on dts/arm/nxp/nxp_imx94x.dtsi:
```diff
@@ -144,6 +145,22 @@
 
+		flexio1: flexio@425c0000 {
+			compatible = "nxp,flexio";
+			reg = <0x425c0000 DT_SIZE_K(4)>;
```
> The commit description says that you enable edma4 but the code does more than that. E.g enable flexio1 etc

The flexio is from nxp_imx943_m33.dts, so i move them from nxp_imx943_m33.dts into the common file nxp_imx94x.dtsi

biwenli commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -1,12 +1,12 @@
 /*
```
> I'm sorry but I don't understand the commit message.
> 
> What does this mean `mcux-sdk-ng/drivers/irqsteer_1 just likes a native irqsteer driver`?
> 
> From my understanding the commit message should look like this:
> 
> ```
> drivers: intc: irqsteer: Drop calling into NXP HAL 
> 
> Supporting irqsteer using NXP HAL becomes increasingly harder with new
> boards.
> 
> For example now there are two incompatible HAL drivers for IRQ steer
> (mcux-sdk-ng/drivers/irqsteer and mcux-sdk-ng/drivers/irqsteer_1).
> 
> In order to avoid overcomplicating code and better scaling code for newer
> boards just drop using the NXP HAL and implement an IRQ Steer native
> Zephyr driver
> ```

That means that the mcux-sdk-ng/drivers/irqsteer_1 looks like a native irqsteer driver.

Sure, will follow the message.

biwenli commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -280,6 +291,12 @@ LOG_MODULE_REGISTER(nxp_irqstr);
+	/* There is one output irq for each group of 64 inputs. */
+	int num_masters;
+	/* One register bit map can represent 32 input interrupts. */
+	int reg_num;
```
> AFAIK IRQSTEER is an configurable IP at hardware integration time so are you sure the groups are always 64? Not clear what is reg_num.

Yes, i'm sure. I have checked all of platforms for the input interrputs per group.
<img width="805" height="296" alt="image" src="https://github.com/user-attachments/assets/b5ab6fd0-185f-4bc6-9401-61b9e5d3c8c8" />

biwenli committed:
boards: nxp: imx943_evk: m33: enable edma4

Enable edma4 by default for lpuart8 of m33(in NETCMIX)
of imx943_evk

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -602,65 +788,82 @@ static void irqsteer_isr_dispatcher(const void *data)
+		.regmap_size = DT_INST_REG_SIZE(n), \
+		.irq_offset = DT_INST_PROP_OR(n, nxp_irq_offset, 0), \
+		.irqs_num = DT_INST_PROP(n, nxp_num_irqs), \
+		.num_masters = DT_INST_PROP(n, nxp_num_irqs) / 64, \
```
> if the `nxp_num_irqs` is `5 * 32`, master counter is 3 (imx8ml dsp), then this won't work.

Yes, you're right. Let's correct it.

biwenli committed:
tests: drivers: uart_async_api: enable it for m33 of imx943_evk

This commit enables tests/driver/uart/async_api demo
for m33 in NETCMIX of imx943_evk.
- build command as follows,
  west build -p always -b imx943_evk/mimx94398/m33
  tests/drivers/uart/uart_async_api

Signed-off-by: Biwen Li <biwen.li@nxp.com>

dbaluta commented:
Tested with SOF on i.MX8MP. Looks good! Thanks @biwenli 

Please check the twister errors.

biwenli commented:
> Tested with SOF on i.MX8MP. Looks good! Thanks @biwenli
> 
> Please check the twister errors.

Okay, got it. Thanks.

biwenli committed:
tests: bluetooth: audio: ascs: fix build issues

Fix build issues,
- error: format %x expects argument of type
  unsigned int, but argument 7 has type ssize_t
  {aka long int} [-Werror=format=]
  zassert_false(ret < 0, "attr->read returned unexpected (err 0x%02x)"
  , BT_GATT_ERR(ret));

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli commented:
Hi @nashif 

 soc/nxp/imx/imx9/imx943/Kconfig.defconfig.mimx94398.m33:24: error: couldn't parse 'configdefault 1ST_LEVEL_INTERRUPT_BITS': configdefault can only contain `default`
Error: See https://docs.zephyrproject.org/latest/build/kconfig/tips.html for more details.

 soc/nxp/imx/imx9/imx943/Kconfig.defconfig.mimx94398.m33:24: error: couldn't parse 'configdefault 1ST_LEVEL_INTERRUPT_BITS': configdefault can only contain `default`

Do you have idea to fix this issue?

biwenli commented:
> Hi @nashif
> 
> soc/nxp/imx/imx9/imx943/Kconfig.defconfig.mimx94398.m33:24: error: couldn't parse 'configdefault 1ST_LEVEL_INTERRUPT_BITS': configdefault can only contain `default` Error: See https://docs.zephyrproject.org/latest/build/kconfig/tips.html for more details.
> 
> soc/nxp/imx/imx9/imx943/Kconfig.defconfig.mimx94398.m33:24: error: couldn't parse 'configdefault 1ST_LEVEL_INTERRUPT_BITS': configdefault can only contain `default`
> 
> Do you have idea to fix this issue?

Fixed.

biwenli committed:
firmware: scmi: fix build issue

Fix build issue:
- error: CONFIG_PM_LOG_LEVEL undeclared here (not in a function)

Signed-off-by: Biwen Li <biwen.li@nxp.com>

ZhaoxiangJin commented:
you need to check [sonarqubecloud](https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=99187&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) reported issues and fix them as much as possible.
